README.md
# DaChoppa.jl

RUN! GO! GET TO DA mixed-integer convex optimization solver, DaChoppa.jl


NEWS.md
DaChoppa release notes
======================


test/nlptest.jl
#  Copyright 2017, Chris Coey and Miles Lubin
#  Copyright 2016, Los Alamos National Laboratory, LANS LLC.
#  This Source Code Form is subject to the terms of the Mozilla Public
#  License, v. 2.0. If a copy of the MPL was not distributed with this
#  file, You can obtain one at http://mozilla.org/MPL/2.0/.

#=========================================================
 NLP model unit tests
=========================================================#

# take a JuMP model and solve, redirecting output
function solve_jump(testname, m, redirect)
    flush(STDOUT)
    flush(STDERR)
    @printf "%-30s... " testname
    tic()

    if redirect
        mktemp() do path,io
            out = STDOUT
            err = STDERR
            redirect_stdout(io)
            redirect_stderr(io)

            status = try
                solve(m)
            catch e
                e
            end

            flush(io)
            redirect_stdout(out)
            redirect_stderr(err)
        end
    else
        status = try
            solve(m)
        catch e
            e
        end
    end
    flush(STDOUT)
    flush(STDERR)

    if isa(status, ErrorException)
        @printf ":%-16s %5.2f s\n" "ErrorException" toq()
    else
        @printf ":%-16s %5.2f s\n" status toq()
    end

    flush(STDOUT)
    flush(STDERR)

    return status
end

# quadratic program tests
function run_qp(mip_solver, log_level, redirect)
    solver=DaChoppaSolver(timeout=120., mip_solver=mip_solver, log_level=(redirect ? 0 : 3))

    testname = "QP optimal"
    @testset "$testname" begin
        m = Model(solver=solver)

        @variable(m, 0 <= x <= 10/3, Int)
        @variable(m, y >= 0)
        @variable(m, 0 <= u <= 10, Int)
        @variable(m, w == 1)

        @objective(m, Min, -3x - y)

        @constraint(m, 3x + 2y <= 18)
        @constraint(m, y^2 <= u*w)

        status = solve_jump(testname, m, redirect)

        @test status == :Optimal
        @test isapprox(getobjectivevalue(m), -12.162277, atol=TOL)
        @test isapprox(getobjbound(m), -12.162277, atol=TOL)
        @test isapprox(getvalue(x), 3, atol=TOL)
        @test isapprox(getvalue(y), 3.162277, atol=TOL)
    end

    testname = "QP maximize"
    @testset "$testname" begin
        m = Model(solver=solver)

        @variable(m, x >= 0, Int)
        @variable(m, y >= 0)
        @variable(m, 0 <= u <= 10, Int)
        @variable(m, w == 1)

        @objective(m, Max, 3x + y)

        @constraint(m, 3x + 2y + 10 <= 20)
        @constraint(m, x^2 <= u*w)

        status = solve_jump(testname, m, redirect)

        @test status == :Optimal
        @test isapprox(getobjectivevalue(m), 9.5, atol=TOL)
        @test isapprox(getobjbound(m), 9.5, atol=TOL)
        @test isapprox(getvalue(x), 3, atol=TOL)
        @test isapprox(getvalue(y), 0.5, atol=TOL)
    end
end

# NLP model tests
function run_nlp(mip_solver, log_level, redirect)
    solver=DaChoppaSolver(timeout=120., mip_solver=mip_solver, log_level=(redirect ? 0 : 3))

    testname = "Nonconvex error"
    @testset "$testname" begin
        m = Model(solver=solver)

        @variable(m, x >= 0, start = 1, Int)
        @variable(m, y >= 0, start = 1)

        @objective(m, Min, -3x - y)

        @constraint(m, 3x + 2y + 10 <= 20)
        @NLconstraint(m, 8 <= x^2 <= 10)

        status = solve_jump(testname, m, redirect)

        @test isa(status, ErrorException)
    end

    testname = "Optimal"
    @testset "$testname" begin
        m = Model(solver=solver)

        @variable(m, x >= 0, start = 1, Int)
        @variable(m, y >= 0, start = 1)

        @objective(m, Min, -3x - y)

        @constraint(m, 3x + 2y + 10 <= 20)
        @constraint(m, x >= 1)
        @NLconstraint(m, x^2 <= 5)
        @NLconstraint(m, exp(y) + x <= 7)

        status = solve_jump(testname, m, redirect)

        @test status == :Optimal
        @test isapprox(getvalue(x), 2.0)
    end

    testname = "Infeasible 1"
    @testset "$testname" begin
        m = Model(solver=solver)

        @variable(m, x >= 0, start = 1, Int)
        @variable(m, y >= 0, start = 1)

        @objective(m, Min, -3x - y)

        @constraint(m, 3x + 2y + 10 <= 20)
        @NLconstraint(m, x^2 >= 9)
        @NLconstraint(m, exp(y) + x <= 2)

        status = solve_jump(testname, m, redirect)

        @test status == :Infeasible
    end

    testname = "Infeasible 2"
    @testset "$testname" begin
        m = Model(solver=solver)

        @variable(m, x >= 0, start = 1, Int)
        @variable(m, y >= 0, start = 1)

        @objective(m, Min, -3x - y)

        @constraint(m, 3x + 2y + 10 <= 20)
        @constraint(m, 6x + 5y >= 30)
        @NLconstraint(m, x^2 >= 8)
        @NLconstraint(m, exp(y) + x <= 7)

        status = solve_jump(testname, m, redirect)

        @test status == :Infeasible
    end

    # testname = "Continuous error"
    # @testset "$testname" begin
    #     m = Model(solver=solver)
    #
    #     @variable(m, x >= 0, start = 1)
    #     @variable(m, y >= 0, start = 1)
    #
    #     @objective(m, Min, -3x - y)
    #
    #     @constraint(m, 3x + 2y + 10 <= 20)
    #     @constraint(m, x >= 1)
    #
    #     @NLconstraint(m, x^2 <= 5)
    #     @NLconstraint(m, exp(y) + x <= 7)
    #
    #     status = solve_jump(testname, m, redirect)
    #
    #     @test isa(status, ErrorException)
    # end

    testname = "Maximization"
    @testset "$testname" begin
        m = Model(solver=solver)

        @variable(m, x >= 0, start = 1, Int)
        @variable(m, y >= 0, start = 1)

        @objective(m, Max, 3x + y)

        @constraint(m, 3x + 2y + 10 <= 20)
        @NLconstraint(m, x^2 <= 9)

        status = solve_jump(testname, m, redirect)

        @test status == :Optimal
        @test isapprox(getobjectivevalue(m), 9.5, atol=TOL)
    end

    # testname = "Nonlinear objective"
    # @testset "$testname" begin
    #     m = Model(solver=solver)
    #
    #     @variable(m, x >= 0, start = 1, Int)
    #     @variable(m, y >= 0, start = 1)
    #
    #     @objective(m, Max, -x^2 - y)
    #
    #     @constraint(m, x + 2y >= 4)
    #     @NLconstraint(m, x^2 <= 9)
    #
    #     status = solve_jump(testname, m, redirect)
    #
    #     @test status == :Optimal
    #     @test isapprox(getobjectivevalue(m), -2.0, atol=TOL)
    #     @test isapprox(getobjbound(m), -2.0, atol=TOL)
    # end
end


test/runtests.jl
#  Copyright 2017, Chris Coey and Miles Lubin
#  Copyright 2016, Los Alamos National Laboratory, LANS LLC.
#  This Source Code Form is subject to the terms of the Mozilla Public
#  License, v. 2.0. If a copy of the MPL was not distributed with this
#  file, You can obtain one at http://mozilla.org/MPL/2.0/.

#=========================================================
 DaChoppa solver unit tests
=========================================================#

using JuMP
import ConicBenchmarkUtilities
using DaChoppa
using Base.Test

include("nlptest.jl")
include("conictest.jl")

# test absolute tolerance and DaChoppa printing level
TOL = 1e-3
ll = 2
redirect = true

# use JuMP list of available solvers
include(Pkg.dir("JuMP", "test", "solvers.jl"))

# MIP solvers
tol_int = 1e-9
tol_feas = 1e-7
tol_gap = 0.0

mip_solvers = Dict{String,MathProgBase.AbstractMathProgSolver}()
if glp
    mip_solvers["GLPK"] = GLPKMathProgInterface.GLPKSolverMIP(msg_lev=GLPK.MSG_OFF, tol_int=tol_int, tol_bnd=tol_feas, mip_gap=tol_gap)
end
if cpx
    mip_solvers["CPLEX"] = CPLEX.CplexSolver(CPX_PARAM_SCRIND=0, CPX_PARAM_EPINT=tol_int, CPX_PARAM_EPRHS=tol_feas, CPX_PARAM_EPGAP=tol_gap)
end
if grb
    mip_solvers["Gurobi"] = Gurobi.GurobiSolver(OutputFlag=0, IntFeasTol=tol_int, FeasibilityTol=tol_feas, MIPGap=tol_gap)
end

# print solvers
println("\nMILP solvers:")
for (i, sname) in enumerate(keys(mip_solvers))
    @printf "%2d  %s\n" i sname
end

# run tests
@testset "MILP solver - $mipname" for (mipname, mip) in mip_solvers
    @testset "NLP models" begin
        println("\nNLP models")
        run_qp(mip, ll, redirect)
        run_nlp(mip, ll, redirect)
    end
    @testset "Exp+SOC models" begin
        println("\nExp+SOC models")
        run_soc(mip, ll, redirect)
        run_expsoc(mip, ll, redirect)
    end
    println()
    flush(STDOUT)
    flush(STDERR)
end
println()


test/conictest.jl
#  Copyright 2017, Chris Coey and Miles Lubin
#  Copyright 2016, Los Alamos National Laboratory, LANS LLC.
#  This Source Code Form is subject to the terms of the Mozilla Public
#  License, v. 2.0. If a copy of the MPL was not distributed with this
#  file, You can obtain one at http://mozilla.org/MPL/2.0/.

#=========================================================
 conic model unit tests
=========================================================#

# take a CBF (conic benchmark format) file and solve, redirecting output
function solve_cbf(testname, probname, solver, redirect)
    flush(STDOUT)
    flush(STDERR)
    @printf "%-30s... " testname
    tic()

    dat = ConicBenchmarkUtilities.readcbfdata("cbf/$(probname).cbf")
    (c, A, b, con_cones, var_cones, vartypes, sense, objoffset) = ConicBenchmarkUtilities.cbftompb(dat)
    if sense == :Max
        c = -c
    end
    flush(STDOUT)
    flush(STDERR)

    m = MathProgBase.ConicModel(solver)

    if redirect
        mktemp() do path,io
            out = STDOUT
            err = STDERR
            redirect_stdout(io)
            redirect_stderr(io)

            MathProgBase.loadproblem!(m, c, A, b, con_cones, var_cones)
            MathProgBase.setvartype!(m, vartypes)
            MathProgBase.optimize!(m)

            flush(io)
            redirect_stdout(out)
            redirect_stderr(err)
        end
    else
        MathProgBase.loadproblem!(m, c, A, b, con_cones, var_cones)
        MathProgBase.setvartype!(m, vartypes)
        MathProgBase.optimize!(m)
    end
    flush(STDOUT)
    flush(STDERR)

    status = MathProgBase.status(m)
    time = MathProgBase.getsolvetime(m)
    if sense == :Max
        objval = -MathProgBase.getobjval(m)
        objbound = -MathProgBase.getobjbound(m)
    else
        objval = MathProgBase.getobjval(m)
        objbound = MathProgBase.getobjbound(m)
    end
    sol = MathProgBase.getsolution(m)
    @printf ":%-16s %5.2f s\n" status toq()
    flush(STDOUT)
    flush(STDERR)

    return (status, time, objval, objbound, sol)
end

# second-order cone model tests
function run_soc(mip_solver, log_level, redirect)
    testname = "SOC optimal"
    probname = "soc_optimal"
    @testset "$testname" begin
        solver = DaChoppaSolver(timeout=120., mip_solver=mip_solver, log_level=log_level)

        (status, time, objval, objbound, sol) = solve_cbf(testname, probname, solver, redirect)

        @test status == :Optimal
        @test isapprox(sol[1], 3, atol=TOL)
        @test isapprox(objval, -9, atol=TOL)
    end

    # testname = "SOC infeasible"
    # probname = "soc_infeasible"
    # @testset "$testname" begin
    #     solver = DaChoppaSolver(timeout=120., mip_solver=mip_solver, log_level=log_level)
    #
    #     (status, time, objval, objbound, sol) = solve_cbf(testname, probname, solver, redirect)
    #
    #     @test status == :Infeasible
    # end

    # testname = "SOCRot optimal"
    # probname = "socrot_optimal"
    # @testset "$testname" begin
    #     solver = DaChoppaSolver(timeout=120., mip_solver=mip_solver, log_level=log_level)
    #
    #     (status, time, objval, objbound, sol) = solve_cbf(testname, probname, solver, redirect)
    #
    #     @test status == :Optimal
    #     @test isapprox(objval, -9, atol=TOL)
    #     @test isapprox(objbound, -9, atol=TOL)
    #     @test isapprox(sol, [1.5, 3, 3, 3], atol=TOL)
    # end

    testname = "SOCRot infeasible"
    probname = "socrot_infeasible"
    @testset "$testname" begin
        solver = DaChoppaSolver(timeout=120., mip_solver=mip_solver, log_level=log_level)

        (status, time, objval, objbound, sol) = solve_cbf(testname, probname, solver, redirect)

        @test status == :Infeasible
    end

    # testname = "Equality constraint"
    # probname = "soc_equality"
    # @testset "$testname" begin
    #     solver = DaChoppaSolver(timeout=120., mip_solver=mip_solver, log_level=log_level)
    #
    #     (status, time, objval, objbound, sol) = solve_cbf(testname, probname, solver, redirect)
    #
    #     @test status == :Optimal
    #     @test isapprox(objval, -sqrt(2), atol=TOL)
    #     @test isapprox(objbound, -sqrt(2), atol=TOL)
    #     @test isapprox(sol, [1, 1/sqrt(2), 1/sqrt(2)], atol=TOL)
    # end
    #
    # testname = "Zero cones"
    # probname = "soc_zero"
    # @testset "$testname" begin
    #     solver = DaChoppaSolver(timeout=120., mip_solver=mip_solver, log_level=log_level)
    #
    #     (status, time, objval, objbound, sol) = solve_cbf(testname, probname, solver, redirect)
    #
    #     @test status == :Optimal
    #     @test isapprox(objval, -sqrt(2), atol=TOL)
    #     @test isapprox(objbound, -sqrt(2), atol=TOL)
    #     @test isapprox(sol, [1, 1/sqrt(2), 1/sqrt(2), 0, 0], atol=TOL)
    # end

    testname = "SOC infeasible binary"
    probname = "soc_infeasible2"
    @testset "$testname" begin
        solver = DaChoppaSolver(timeout=120., mip_solver=mip_solver, log_level=log_level)

        (status, time, objval, objbound, sol) = solve_cbf(testname, probname, solver, redirect)

        @test status == :Infeasible
    end
end

# exponential and second-order cone tests
function run_expsoc(mip_solver, log_level, redirect)
    testname = "Exp optimal"
    probname = "exp_optimal"
    @testset "$testname" begin
        solver = DaChoppaSolver(timeout=120., mip_solver=mip_solver, log_level=log_level)

        (status, time, objval, objbound, sol) = solve_cbf(testname, probname, solver, redirect)

        @test status == :Optimal
        @test isapprox(objval, -8, atol=TOL)
        @test isapprox(objbound, -8, atol=TOL)
        @test isapprox(sol[1:2], [2, 2], atol=TOL)
    end

    testname = "ExpSOC optimal"
    probname = "expsoc_optimal"
    @testset "$testname" begin
        solver = DaChoppaSolver(timeout=120., mip_solver=mip_solver, log_level=log_level)

        (status, time, objval, objbound, sol) = solve_cbf(testname, probname, solver, redirect)

        @test status == :Optimal
        @test isapprox(objval, -7.609438, atol=TOL)
        @test isapprox(objbound, -7.609438, atol=TOL)
        @test isapprox(sol[1:2], [2, 1.609438], atol=TOL)
    end

    # testname = "ExpSOC optimal 3"
    # probname = "expsoc_optimal3"
    # @testset "$testname" begin
    #     solver = DaChoppaSolver(timeout=120., mip_solver=mip_solver, log_level=log_level)
    #
    #     (status, time, objval, objbound, sol) = solve_cbf(testname, probname, solver, redirect)
    #
    #     @test status == :Optimal
    #     @test isapprox(objval, -7, atol=TOL)
    #     @test isapprox(objbound, -7, atol=TOL)
    #     @test isapprox(sol[1:2], [1, 2], atol=TOL)
    # end

    testname = "Exp large (gatesizing)"
    probname = "exp_gatesizing"
    @testset "$testname" begin
        solver = DaChoppaSolver(timeout=120., mip_solver=mip_solver, log_level=log_level)

        (status, time, objval, objbound, sol) = solve_cbf(testname, probname, solver, redirect)

        @test status == :Optimal
        @test isapprox(objval, 8.33333, atol=TOL)
        @test isapprox(objbound, 8.33333, atol=TOL)
        @test isapprox(exp.(sol[1:7]), [2, 3, 3, 3, 2, 3, 3], atol=TOL)
    end

    testname = "Exp large 2 (Ising)"
    probname = "exp_ising"
    @testset "$testname" begin
        solver = DaChoppaSolver(timeout=120., mip_solver=mip_solver, log_level=log_level)

        (status, time, objval, objbound, sol) = solve_cbf(testname, probname, solver, redirect)

        @test status == :Optimal
        @test isapprox(objval, 0.696499, atol=TOL)
        @test isapprox(objbound, 0.696499, atol=TOL)
        @test isapprox(sol[end-8:end], [0, 0, 1, 0, 0, 0, 2, 1, 0], atol=TOL)
    end
end


src/solver.jl

export DaChoppaSolver

# Dummy solver
type UnsetSolver <: MathProgBase.AbstractMathProgSolver end

# DaChoppa solver
type DaChoppaSolver <: MathProgBase.AbstractMathProgSolver
    log_level::Int              # Verbosity flag
    timeout::Float64            # Time limit (in seconds)
    rel_gap::Float64            # Relative optimality gap
    zero_tol::Float64           # Zero tolerance for cut coefficients
    feas_tol::Float64           # Absolute feasibility tolerance
    mip_solver::MathProgBase.AbstractMathProgSolver # MIP solver
end

function DaChoppaSolver(;
    log_level = 1,
    timeout = Inf,
    rel_gap = 1e-5,
    zero_tol = 1e-10,
    feas_tol = 1e-6,
    mip_solver = UnsetSolver(),
    )

    if mip_solver == UnsetSolver()
        error("No MIP solver specified (set mip_solver)\n")
    end

    DaChoppaSolver(log_level, timeout, rel_gap, zero_tol, feas_tol, deepcopy(mip_solver))
end

# Create DaChoppa conic model
MathProgBase.ConicModel(s::DaChoppaSolver) = MathProgBase.ConicModel(ConicNonlinearBridge.ConicNLPWrapper(nlp_solver=s))

# Create DaChoppa LinearQuadratic model
MathProgBase.LinearQuadraticModel(s::DaChoppaSolver) = MathProgBase.NonlinearToLPQPBridge(MathProgBase.NonlinearModel(s))

# Create DaChoppa nonlinear model
MathProgBase.NonlinearModel(s::DaChoppaSolver) = DaChoppaNonlinearModel(s.log_level, s.timeout, s.rel_gap, s.zero_tol, s.feas_tol, s.mip_solver)


src/algorithm.jl

#=========================================================
 Nonlinear model object
=========================================================#

type DaChoppaNonlinearModel <: MathProgBase.AbstractNonlinearModel
    log_level::Int              # Verbosity flag
    timeout::Float64            # Time limit (in seconds)
    rel_gap::Float64            # Relative optimality gap
    zero_tol::Float64           # Zero tolerance for cut coefficients
    feas_tol::Float64           # Absolute feasibility tolerance
    mip_solver::MathProgBase.AbstractMathProgSolver # MIP solver

    sense::Symbol
    numvar::Int
    vartypes::Vector{Symbol}
    warmstart::Vector{Float64}
    numconstr::Int
    geqnlcons::Vector{Int}
    leqnlcons::Vector{Int}
    conlbs::Vector{Float64}
    conubs::Vector{Float64}
    conjac::Vector{Vector{Int}}
    d

    status::Symbol
    oam::Model
    x::Vector{JuMP.Variable}
    incumbent::Vector{Float64}
    objval::Float64
    objbound::Float64
    objgap::Float64
    totaltime::Float64

    function DaChoppaNonlinearModel(log_level, timeout, rel_gap, zero_tol, feas_tol, mip_solver)
        m = new()

        m.log_level = log_level
        m.timeout = timeout
        m.rel_gap = rel_gap
        m.zero_tol = zero_tol
        m.feas_tol = feas_tol
        m.mip_solver = mip_solver

        m.status = :NotLoaded
        m.incumbent = Float64[]
        m.objval = Inf
        m.objbound = -Inf
        m.objgap = Inf
        m.totaltime = 0.0

        return m
    end
end


#=========================================================
 MathProgBase functions
=========================================================#

MathProgBase.numvar(m::DaChoppaNonlinearModel) = m.numvar
MathProgBase.numconstr(m::DaChoppaNonlinearModel) = m.numconstr
MathProgBase.setwarmstart!(m::DaChoppaNonlinearModel, x) = (m.incumbent = x) # TODO set values on mip model, maybe add cuts at point
MathProgBase.setvartype!(m::DaChoppaNonlinearModel, v::Vector{Symbol}) = (m.vartypes = v)

function MathProgBase.loadproblem!(m::DaChoppaNonlinearModel, numvar, numconstr, l, u, lb, ub, sense, d)
    m.numvar = numvar
    m.numconstr = numconstr
    m.sense = sense
    m.incumbent = fill(NaN, numvar)

    MathProgBase.initialize(d, [:Jac])
    m.d = d
    xzero = zeros(m.numvar)

    # Initialize JuMP model for MIP OA
    m.oam = oam = JuMP.Model(solver=m.mip_solver)
    m.x = x = @variable(oam, [j in 1:numvar], lowerbound=l[j], upperbound=u[j])

    # Set objective
    if !MathProgBase.isobjlinear(d)
        # TODO allow quadratic obj, check isobjquadratic and use obj_expr
        error("Objective function must be linear")
    end
    # TODO no constant in obj
    @assert MathProgBase.eval_f(d, xzero) == 0.0
    c = zeros(m.numvar)
    MathProgBase.eval_grad_f(d, c, xzero)
    if sense == :Max
        c .*= -1
    end
    @objective(oam, Min, dot(c, x))

    # Add constraints
    isconlin = fill(false, numconstr)
    numlin = 0
    contolin = fill(-1, numconstr)
    leqnlcons = Int[]
    geqnlcons = Int[]
    for i in 1:numconstr
        if MathProgBase.isconstrlinear(d, i)
            isconlin[i] = true
            numlin += 1
            contolin[i] = numlin
        elseif isfinite(lb[i]) && isfinite(ub[i])
            error("Nonlinear equality/two-sided constraints not accepted")
        elseif isfinite(lb[i])
            push!(geqnlcons, i)
        else
            @assert isfinite(ub[i])
            push!(leqnlcons, i)
        end
    end
    m.geqnlcons = geqnlcons
    m.leqnlcons = leqnlcons
    m.conlbs = lb
    m.conubs = ub

    (jac_I, jac_J) = MathProgBase.jac_structure(d)
    jac_V = zeros(length(jac_I))
    MathProgBase.eval_jac_g(d, jac_V, xzero)
    A_lin_I = Int[]
    A_lin_J = Int[]
    A_lin_V = Float64[]
    conjac = [Int[] for i in 1:numconstr]
    for k in 1:length(jac_I)
        i = jac_I[k]
        if isconlin[i]
            push!(A_lin_I, contolin[i])
            push!(A_lin_J, jac_J[k])
            push!(A_lin_V, jac_V[k])
        else
            push!(conjac[i], k)
        end
    end
    m.conjac = conjac
    A_lin = sparse(A_lin_I, A_lin_J, A_lin_V, numlin, numvar)
    convals = zeros(numconstr)
    MathProgBase.eval_g(d, convals, xzero)
    b_lin = convals[isconlin] - A_lin*xzero
    lb_lin = lb[isconlin]
    ub_lin = ub[isconlin]

    @constraint(oam, lb_lin .<= A_lin*x + b_lin .<= ub_lin)

    m.status = :Loaded
end

function MathProgBase.optimize!(m::DaChoppaNonlinearModel)
    totaltime = time()
    m.status = :Unsolved

    for j in 1:m.numvar
        setcategory(m.x[j], m.vartypes[j])
    end

    conoa = zeros(m.numconstr)
    (jac_I, jac_J) = MathProgBase.jac_structure(m.d)
    jac_V = zeros(length(jac_I))
    num_cuts = 0
    num_cbs = 0

    function callback_lazy(cb)
        num_cbs += 1
        xoa = getvalue(m.x)
        MathProgBase.eval_g(m.d, conoa, xoa)
        MathProgBase.eval_jac_g(m.d, jac_V, xoa)
        # @show xoa
        # @show conoa
        # @show jac_V

        # Add gradient cuts on nonlinear constraints
        for i in m.geqnlcons
            if m.conlbs[i] - conoa[i] > m.feas_tol
                # @show m.conlbs[i] - conoa[i]
                # @show sum(jac_V[k]*(m.x[jac_J[k]] - xoa[jac_J[k]]) for k in m.conjac[i])
                @lazyconstraint(cb, m.conlbs[i] <= conoa[i] + sum(jac_V[k]*(m.x[jac_J[k]] - xoa[jac_J[k]]) for k in m.conjac[i]))
                num_cuts += 1
            end
        end
        for i in m.leqnlcons
            if conoa[i] - m.conubs[i] > m.feas_tol
                # @show conoa[i] - m.conubs[i]
                # @show sum(jac_V[k]*(m.x[jac_J[k]] - xoa[jac_J[k]]) for k in m.conjac[i])
                @lazyconstraint(cb, conoa[i] + sum(jac_V[k]*(m.x[jac_J[k]] - xoa[jac_J[k]]) for k in m.conjac[i]) <= m.conubs[i])
                num_cuts += 1
            end
        end
    end
    addlazycallback(m.oam, callback_lazy)

    # Start MIP solver
    if isfinite(m.timeout) && applicable(MathProgBase.setparameters!, m.mip_solver)
        MathProgBase.setparameters!(m.mip_solver, TimeLimit=max(0.1, m.timeout - (time() - totaltime)))
        setsolver(m.oam, m.mip_solver)
    end
    if m.log_level > 0
        println("\nDaChoppa solver starting...")
    end
    mip_time = time()
    status_mip = solve(m.oam, suppress_warnings=true)
    mip_time = time() - mip_time

    if status_mip in (:Optimal, :Suboptimal)
        m.incumbent = getvalue(m.x)
        m.objval = getobjectivevalue(m.oam)
        m.objbound = MathProgBase.getobjbound(m.oam)
        @assert isfinite(m.objval) && isfinite(m.objbound)
        m.objgap = (m.objval - m.objbound)/(abs(m.objval) + 1e-5)
        m.status = status_mip
    elseif status_mip == :UserLimit
        m.objval = getobjectivevalue(m.oam)
        m.objbound = MathProgBase.getobjbound(m.oam)
        if isfinite(m.objval)
            m.incumbent = getvalue(m.x)
            if isfinite(m.objbound)
                m.objgap = (m.objval - m.objbound)/(abs(m.objval) + 1e-5)
            end
        end
        m.status = status_mip
    elseif status_mip == :Infeasible
        m.status = status_mip
    elseif status_mip in (:Unbounded, :InfeasibleOrUnbounded)
        warn("MIP solver status was $status_mip (try using initial outer approximation cuts)")
        m.status = :UnboundedRelaxation
    else
        warn("MIP solver status was $status_mip")
        m.status = :MIPFailure
    end

    m.totaltime = time() - totaltime
    if m.log_level > 0
        println("\nDaChoppa solver finished...")
        @printf "Status           %13s\n" m.status
        if isfinite(m.objgap)
            @printf "Objective value  %13.5f\n" (m.sense == :Min) ? m.objval : -m.objval
            @printf "Objective bound  %13.5f\n" (m.sense == :Min) ? m.objbound : -m.objbound
            @printf "Objective gap    %13.5f\n" (m.sense == :Min) ? m.objgap : -m.objgap
        end
        @printf "Total time       %13.5f s\n" m.totaltime
        @printf "MIP time         %13.5f s\n" mip_time
        @printf "Cuts added       %13d\n" num_cuts
        @printf "Callbacks        %13d\n" num_cbs
        println()
    end
end

MathProgBase.status(m::DaChoppaNonlinearModel) = m.status
MathProgBase.getobjval(m::DaChoppaNonlinearModel) = (m.sense == :Min) ? m.objval : -m.objval
MathProgBase.getobjbound(m::DaChoppaNonlinearModel) = (m.sense == :Min) ? m.objbound : -m.objbound
MathProgBase.getobjgap(m::DaChoppaNonlinearModel) = (m.sense == :Min) ? m.objgap : -m.objgap
MathProgBase.getsolution(m::DaChoppaNonlinearModel) = m.incumbent
MathProgBase.getsolvetime(m::DaChoppaNonlinearModel) = m.totaltime
MathProgBase.getnodecount(m::DaChoppaNonlinearModel) = MathProgBase.getnodecount(m.oam)


examples/cardls.jl
# Cardinality constrained least squares
# m sample points in R^d, in matrix A in R^(m,d)
# m measurements, in vector b
# Estimate x in R^d to minimize ||Ax-b||_2
# Constrain at most k components of x to be nonzero (select k features only)
#   minimize    1/2*||Ax-b||_2^2 + 1/2*rho*||x||_2^2
#   subject to  ||x||_0 <= k

using JuMP, DaChoppa

# Set up QP JuMP model, solve, print solution
# xB is a bound on the absolute values of the estimate variables x_j
function miqp_cardls(m, d, A, b, k, rho, xB, solver)
    mod = Model(solver=solver)
    @variable(mod, x[1:d])
    @variable(mod, z[1:d], Bin)
    @variable(mod, u >= 0)
    @variable(mod, v >= 0)
    @objective(mod, Min, u + rho*v)
    @variable(mod, t[1:m])
    @constraint(mod, t .== A*x - b)
    @variable(mod, w == 2)
    @constraint(mod, sum(t.^2) <= u*w)
    @constraint(mod, sum(x.^2) <= v*w)
    @constraint(mod, x .<= xB.*z)
    @constraint(mod, x .>= -xB.*z)
    @constraint(mod, sum(z) <= k)

    solve(mod)
    println("Selected features (z) = \n$(getvalue(z))\n")
end


#=========================================================
Choose solvers and options
=========================================================#

rel_gap = 1e-5


# using Cbc
# mip_solver = CbcSolver()

using CPLEX
mip_solver = CplexSolver(
    CPX_PARAM_SCRIND=1,
    CPX_PARAM_EPINT=1e-8,
    CPX_PARAM_EPRHS=1e-7,
    CPX_PARAM_EPGAP=rel_gap,
)


solver = DaChoppaSolver(
    log_level=1,
    rel_gap=rel_gap,
	mip_solver=mip_solver,
)


#=========================================================
Specify/generate data
=========================================================#

d = 6 # Dimension of feature space
m = 20 # Number of samples

srand(100) # Change or comment random seed to get different data
A = randn(m, d) # Sample point matrix (rows are samples)
b = randn(m) # Sample measurement vector

k = floor(Int, d/2) # Number of features to select (||x||_0 <= k)

rho = 1. # Ridge regularization multiplier

xB = 4 # Bound on absolute values of estimate variables (|x_j| <= xB)


#=========================================================
Solve JuMP model
=========================================================#

miqp_cardls(m, d, A, b, k, rho, xB, solver)


examples/gatesizing.jl
# Electrical circuit gate sizing example adapted from Boyd, Kim, Vandenberghe, and Hassibi, "A Tutorial on Geometric Programming", p28
# CVX code adapted from simple_dig_ckt_sizing.m by Almir Mutapcic
# http://cvxr.com/cvx/examples/gp_tutorial/html/simple_dig_ckt_sizing.html
#
# Choose gate scale factors y_i to minimize ckt delay, subject to limits on the total area and power
#   minimize     D
#   subject to   P <= Pmax,
#                A <= Amax
#                y >= 1
# Where variables are scale factors y, which are constrained to equal discrete values
# Data is hard-coded from the circuit shown in figure 4 of "A Tutorial on Geometric Programming"

using Convex, DaChoppa

# Set up Convex.jl model, solve, print solution
function gatesizing(yUB, solver)
    fe = [1, 0.8, 1, 0.7, 0.7, 0.5, 0.5] .* [1, 2, 1, 1.5, 1.5, 1, 2]
    Cout6 = 10
    Cout7 = 10

    y = Variable(7, Positive())
    z = Variable(yUB, 7, Positive(), :Bin)

    D1 = exp(-y[1]) + exp(-y[1] + y[4])
    D2 = 2 * exp(-y[2]) + exp(-y[2] + y[4]) + exp(-y[2] + y[5])
    D3 = 2 * exp(-y[3]) + exp(-y[3] + y[5]) + exp(-y[3] + y[7])
    D4 = 2 * exp(-y[4]) + exp(-y[4] + y[6]) + exp(-y[4] + y[7])
    D5 = exp(-y[5]) + exp(-y[5] + y[7])
    D6 = Cout6 * exp(-y[6])
    D7 = Cout7 * exp(-y[7])

    P = minimize(
        maximum([(D1+D4+D6), (D1+D4+D7), (D2+D4+D6), (D2+D4+D7), (D2+D5+D7), (D3+D5+D6), (D3+D7)]),
        sum(fe .* exp(y)) <= 20,
        sum(exp(y)) <= 100)

    for i in 1:7
        P.constraints += (sum(z[:,i]) == 1)
        P.constraints += (y[i] == sum([log(j) * z[j,i] for j=1:yUB]))
    end

    solve!(P, solver)

    println("\nCircuit delay (obj) = $(P.optval)")
    println("Scale factors (exp(y)):\n$(exp.(y.value))")
    println("Value indicators (z):\n$(round.(z.value))")
end


#=========================================================
Choose solvers and options
=========================================================#

rel_gap = 1e-5


# using Cbc
# mip_solver = CbcSolver()

using CPLEX
mip_solver = CplexSolver(
    CPX_PARAM_SCRIND=1,
    CPX_PARAM_EPINT=1e-8,
    CPX_PARAM_EPRHS=1e-7,
    CPX_PARAM_EPGAP=rel_gap,
)


solver = DaChoppaSolver(
    log_level=1,
    rel_gap=rel_gap,
	mip_solver=mip_solver,
)


#=========================================================
Specify data
=========================================================#

# Integer upper bound on integer gate size factors
# Optimal solution for this instances has all exp(y) <= 3,
# so yUB larger than 2 is not constraining, but impacts number of variables
yUB = 3


#=========================================================
Solve Convex.jl model
=========================================================#

# Optimal solution is [2,3,3,3,2,3,3] with value 8.3333
gatesizing(yUB, solver)
